### PR TITLE
feature/employee list projects

### PR DESCRIPTION
**Pull request by ncocquio**

### CODE REVIEW

File: src/components/EmployeeList/EmployeeList.tsx [hunk 1]
```diff
@@ -56,14 +60,14 @@ const filterEmployees = (employees: EmployeeModel[], filterString: string): Empl
...
-const EmployeeListItem: Component<EmployeeRow> = ({ employee, contracts, actions }) => {
+const EmployeeListItem: Component<EmployeeRow> = ({ employee, projects, contract, contracts, actions }) => {
   return (
     <tr>
       <td><AvatarItem fullName={employee.fullName} /></td>
       <td>{employee.roleName}</td>
-      <td>{/* #Projects */}</td>
-      <td>{/* Pensum */}</td>
-      <td>{/* Contracts */}</td>
+      <td>{(projects != undefined) ? projects.size : 0}</td>
```
Comment (kelvinlouis) [inline src/components/EmployeeList/EmployeeList.tsx]: projects != null

File: src/components/EmployeeList/EmployeeList.tsx [hunk 2]
```diff
@@ -56,14 +60,14 @@ const filterEmployees = (employees: EmployeeModel[], filterString: string): Empl
...
+      <td>{(contract != undefined) ? `${contract.pensumPercentage}%` : "-"}</td>
```
Comment (kelvinlouis) [inline src/components/EmployeeList/EmployeeList.tsx]: contract != null

File: src/components/EmployeeList/EmployeeList.tsx [hunk 3]
```diff
@@ -96,14 +100,67 @@ const EmployeeListItem: Component<EmployeeRow> = ({ employee, contracts, actions
...
+        employeesProjectMap.set(projectManagerId, new Set())
+      }
+
+      const set = employeesProjectMap.get(projectManagerId)!;
+      set.add(project);
+    }
+  });
+
+  employees.forEach((employee) => {
+    const sortedContracts = contracts.filter(c => c.employeeId == employee.id).sort((a, b) => moment(a.endDate, "YYY-MM-DD").diff(moment(b.endDate, "YYY-MM-DD")))
```
Comment (kelvinlouis) [inline src/components/EmployeeList/EmployeeList.tsx]: 1. use constants for the date format.
2. improve readability by indenting chaining:
3. move sort comparison to a util function

```
const compareByDate = (a: ContractModel, b: ContractModel): number {
  return moment(a.endDate, "YYY-MM-DD").diff(moment(b.endDate, "YYY-MM-DD")
}

contracts
  .filter(c => c.employeeId == employee.id)
  .sort((a, b) => compareByDate(a,b))
```

File: src/components/EmployeeList/EmployeeList.tsx [hunk 4]
```diff
@@ -96,14 +100,67 @@ const EmployeeListItem: Component<EmployeeRow> = ({ employee, contracts, actions
...
+    employeesContractMap.set(employee.id, relevantContract);
+  });
+
   const createEmployeeListItem = (employee: EmployeeModel) => {
     const employeeContracts = contracts!.filter(contract => contract.employeeId === employee.id);
+    const projects = employeesProjectMap.get(employee.id);
+    const relevantContract = employeesContractMap.get(employee.id);
 
-    return <EmployeeListItem key={employee.id} employee={employee} contracts={employeeContracts} actions={actions} />;
+    return <EmployeeListItem key={employee.id} employee={employee} projects={projects} contract={relevantContract} contracts={employeeContracts} actions={actions} />;
```
Comment (kelvinlouis) [inline src/components/EmployeeList/EmployeeList.tsx]: Indentation of properties:
+ Does it need both `contracts` and `contract`?
```
<EmployeeListItem
  key={employee.id}
  employee={employee}
  projects={projects}
  contract={relevantContract}
  contracts={employeeContracts}
  actions={actions}
/>
```

File: src/components/ProjectList/ProjectList.tsx [hunk 5]
```diff
@@ -118,7 +129,7 @@ const ProjectList: Component<Props> = ({ state, actions }) => {
         </tr>
       </thead>
       <tbody>
-        {projects && projects.map((project: ProjectModel) => createProjectRowItem(project))}
+        {filteredProjects.map((project: ProjectModel) => createProjectRowItem(project))}
```
Comment (kelvinlouis) [inline src/components/ProjectList/ProjectList.tsx]: Please also show an indication of how many projects are visible.
Refer to the Employees Table.

File: src/components/EmployeeList/EmployeeList.tsx (hunk 3 above)
Comment (ncocquio) [reply src/components/EmployeeList/EmployeeList.tsx]: even better, as dates are now Moments
```
const sortedContracts = contracts
                              .filter(c => c.employeeId == employee.id)
                              .sort((a, b) => a.endDate.diff(b.endDate))
```

File: src/components/EmployeeList/EmployeeList.tsx (hunk 4 above)
Comment (ncocquio) [reply src/components/EmployeeList/EmployeeList.tsx]: I know, it's not pretty. But computing the contractId and getting it our of contracts in the EmployeeListItem would be even more complicated.

File: src/components/EmployeeList/EmployeeList.tsx [hunk 6]
```diff
@@ -96,14 +100,67 @@ const EmployeeListItem: Component<EmployeeRow> = ({ employee, contracts, actions
...
+    if (sortedContracts.length > 0) {
+      relevantContract = sortedContracts.reduce((prev, curr) => {
+        if (moment().isAfter(curr.endDate)) {
+          return curr;
+        } else {
+          return prev ? prev : curr;
+        }
+      });
+    }
+    employeesContractMap.set(employee.id, relevantContract);
```
Comment (kelvinlouis) [inline src/components/EmployeeList/EmployeeList.tsx]: employeesLatestContractMap

File: src/components/EmployeeList/EmployeeList.tsx (hunk 1 above)
Comment (ncocquio) [reply src/components/EmployeeList/EmployeeList.tsx]: fixed

File: src/components/EmployeeList/EmployeeList.tsx (hunk 2 above)
Comment (ncocquio) [reply src/components/EmployeeList/EmployeeList.tsx]: fixed

File: src/components/ProjectList/ProjectList.tsx (hunk 5 above)
Comment (ncocquio) [reply src/components/ProjectList/ProjectList.tsx]: fixed

File: src/components/EmployeeList/EmployeeList.tsx (hunk 6 above)
Comment (ncocquio) [reply src/components/EmployeeList/EmployeeList.tsx]: changed

**Pull request merged**